src/theory/arith/nl/nl_model.cpp

```diff
@@ -63,46 +57,43 @@ void NlModel::resetCheck()
   d_check_model_solved.clear();
   d_check_model_bounds.clear();
   d_check_model_witnesses.clear();
-  d_check_model_vars.clear();
-  d_check_model_subs.clear();
+  d_substitutions.clear();
 }
 
-Node NlModel::computeConcreteModelValue(Node n)
+Node NlModel::computeConcreteModelValue(TNode n)
 {
   return computeModelValue(n, true);
 }
 
-Node NlModel::computeAbstractModelValue(Node n)
+Node NlModel::computeAbstractModelValue(TNode n)
 {
   return computeModelValue(n, false);
 }
 
-Node NlModel::computeModelValue(Node n, bool isConcrete)
+Node NlModel::computeModelValue(TNode n, bool isConcrete)
 {
-  unsigned index = isConcrete ? 0 : 1;
-  std::map<Node, Node>::iterator it = d_mv[index].find(n);
-  if (it != d_mv[index].end())
+  auto& cache = isConcrete ? d_concreteModelCache : d_abstractModelCache;
+  if (auto it = cache.find(n); it != cache.end())
   {
     return it->second;
   }
-  Trace("nl-ext-mv-debug") << "computeModelValue " << n << ", index=" << index
-                           << std::endl;
+  Trace("nl-ext-mv-debug") << "computeModelValue " << n
+                           << ", isConcrete=" << isConcrete << std::endl;
   Node ret;
-  Kind nk = n.getKind();
   if (n.isConst())
   {
     ret = n;
   }
-  else if (!isConcrete && hasTerm(n))
+  else if (auto it = d_arithVal.find(n); !isConcrete && it != d_arithVal.end())
```

Comment:
This is a bit ugly, can we split this up so that we aren't declaring variables within a conditional?

Reply:
There are two reasons I did it like this:
1. this makes `it` a variable that is local to this `if`.
2. it is actually pretty hard to rewrite this code so that the overall structure (the `if-then-else`) makes sense and at the same time only have the `d_arithVal.find()` executed when necessary.

Reply:
I would prefer a refactoring where e.g. a new method is used, like `bool hasTerm(TNode n, TNode& rep)`.  When this returns true, it sets `rep` to the value of the lookup.

Reply:
I don't know. The usage of such output parameters is usually discouraged (IMHO for good reason). This is standard C++ syntax meant for exactly such cases.

I see that this syntax is rather new (C++17) and as such not widely used in our code base (yet). However, I think it is not hard to understand (consider the resemblance in both syntax and semantics to a `for` loop), and we should really use it when appropriate.